src/libos/src/process/do_vfork.rs

```diff
@@ -0,0 +1,168 @@
+
+    // Store parent and the vforked child
+    let parent_pid = current.process().pid();
+    let mut vfork_process_table = VFORK_PROCESS_TABLE.lock().unwrap();
+    if let Some(_) = vfork_process_table.insert(parent_pid, new_pid) {
+        return_errno!(EINVAL, "current process's vfork has not returned yet");
+    }
+
+    // Save parent's user context in TLS
+    VFORK_CPUCONTEXT.with(|cell| {
+        let mut p_context = cell.borrow_mut();
+        *p_context = unsafe { Some((*context).clone()) };
+    });
+
+    // Save parent's file table
+    let mut vfork_file_tables = VFORK_PARENT_FILE_TABLES.lock().unwrap();
+    let parent_file_table = {
+        let mut current_file_table = current.files().lock().unwrap();
+        let new_file_table = current_file_table.clone();
+        // FileTable contains non-cloned struct, so here we do a memory replacement.
+        mem::replace(&mut *current_file_table, new_file_table)
+    };
+    if let Some(_) = vfork_file_tables.insert(parent_pid, parent_file_table) {
+        return_errno!(EINVAL, "current process's vfork has not returned yet");
+    }
+
+    // This is the first time return and will return as child.
+    // The second time return will return as parent in vfork_return_to_parent.
+    info!("vfork child pid = {:?}", new_pid);
+    return Ok(0 as isize);
+}
+
+// Check if the calling process is a vforked child process that reuse parent's task and pid.
+pub fn is_vforked_child_process() -> bool {
+    // Due to current limitation, the child process is reusing parent process's task and pid.
+    // Thus parent pid is current pid.
+    let parent_pid = current!().process().pid();
+    let vfork_process_table = VFORK_PROCESS_TABLE.lock().unwrap();
+
+    if let Some(_) = vfork_process_table.get(&parent_pid) {
+        return true;
+    } else {
+        return false;
+    }
+}
+
+// Return to parent process to continue executing
+pub fn vfork_return_to_parent(
+    mut context: *mut CpuContext,
+    current_ref: &ThreadRef,
+) -> Result<isize> {
+    return restore_parent_process(context, current_ref);
+}
+
+fn restore_parent_process(mut context: *mut CpuContext, current_ref: &ThreadRef) -> Result<isize> {
+    let current_thread = current!();
+    let current_pid = current_ref.process().pid();
+
+    // Restore the CpuContext
+    VFORK_CPUCONTEXT.with(|cell| {
+        let mut cell = cell.borrow_mut();
+        let previous_context = cell.unwrap();
+        unsafe { *context = previous_context };
+        *cell = None;
+    });
+
+    // Restore parent file table
+    let parent_file_table = {
+        let mut parent_file_tables = VFORK_PARENT_FILE_TABLES.lock().unwrap();
+        if let Some(table) = parent_file_tables.remove(&current_pid) {
+            table
+        } else {
+            return_errno!(EFAULT, "couldn't restore parent file table");
+        }
+    };
+    let mut current_file_table = current_ref.files().lock().unwrap();
+    *current_file_table = parent_file_table;
```

Comment:
Should mem::replace be used?

Reply:
Add more comments. Thanks for your review.